R/db.R

```diff
@@ -27,6 +27,16 @@
 ##'   be validated on open (currently only applicable with \code{type
 ##'   = "destination"}).  This is primarily intended for internal use.
 ##'
+##' @param instance Used only by \code{type = "source"}, and used to
+##'   select the instance, where multiple instances are configured.
+##'   The options here can vary - use a single \emph{unnamed}
+##'   character string to indicate an instance to match.  If given,
+##'   then this name must be present in all databases that support
+##'   instances, and will be ignored by all that do not.  Otherwise,
+##'   provide \emph{named} character vector with the name being the
+##'   source database name and the value being the instance for this
+##'   database (e.g. \code{source = "production"}).
+##'
```

Comment:
Hmm... I am probably a bit rusty on how this works - but I'm finding this para a bit hard to get my head around. It may be easiest to refer to the vignette and have an expanded example, as we probably don't want masses of documentation right here...

I think it's the middle sentence about whether databases support instances that, and what happens if the instance is ignored, that is getting me...

For the VIMC scenario:
So different "databases" might be (eg, source (montagu) and annex?). In that case, annex wouldn't "support" instances (or rather, we wouldn't specify any - because we've only got one of it), but montagu could have instances production, science, uat and latest as instances, in the config file.

So: if we had source (montagu) and annex, we could specify instance="uat", which would do something for montagu, but be safely ignored for the one and only annex.

And if we did have separate instances on annex, then we could specify:
instance = c(source = "uat", annex = "staging_annex")

although then I think I've gone wrong somewhere with the sentence that says this is only used by type="source".

Reply:
Here is the vimc situation: https://github.com/vimc/montagu-reports/compare/vimc-3318 - once this PR is settled I can open a PR for this and get it going.

So we only have one annex instance and the named version of the instance argument would not be used.  It's probably overkill but I think gets us all the way to the most general solution.

Our database instances are going to be `production`, `science`, `uat` and `latest` corresponding to all the full copies of the db.  Annex always refers to whatever is configured on that machine (though we could do this with instances it's a bit awkward as we have one global annex on annex.montagu and a fake annex on uat which is used only by uat).